runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10393,227 +10393,637 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+   generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::vsldoi, node, vtmp2Reg, vconstant0Reg, vtmp2Reg, 8);
+   // mask out the mis-aligned bits in vtmp1Reg
+   if (isLE) {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vslo, node, vtmp2Reg, vconstantNegReg, vtmp2Reg);
+   } else {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vsro, node, vtmp2Reg, vconstantNegReg, vtmp2Reg);
+   }
+   if (elementType == TR::Int32) // we don't need to unpack 32 bit values
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vand, node, low4Reg, vtmp1Reg, vtmp2Reg);
+      }
+   else
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vand, node, vtmp1Reg, vtmp1Reg, vtmp2Reg);
+      }
+
+   // Since we are putting the initial value at the most significant word in our
+   // accumulators, we must multiply it by 31 for every element inserted from the mis-aligned batch
+   if (nonZeroInitial)
+      {
+      // tempReg = (128 - tempReg) = number of bits in the mis-aligned batch
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::nor, node, tempReg, tempReg, tempReg);
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, tempReg, tempReg, 129);
+      // tempReg = tempReg / size * 4 = tempReg >> (lg(size)-2) = the address of the multiplier
+      switch (elementType)
+         {
+         case TR::Int8:
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, tempReg, tempReg, 1);
+            break;
+         case TR::Int16:
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, tempReg, tempReg, 2);
+            break;
+         case TR::Int32:
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, tempReg, tempReg, 3);
+            break;
+         default:
+            TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+         }
+      // adjust for the first 4 elements in the multiplier array
+      if (isLE) {
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, tempReg, tempReg, 0xF);
+      } else { // for BE, we want the correct value to be in the third word
+         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, tempReg, tempReg, 0xF-12);
+      }
+
+      // move the value into a temporary register
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vor, node, vtmp2Reg, initialValueReg, initialValueReg);
+      // multiply
+      generateTrg1MemInstruction(cg, TR::InstOpCode::lxvw4x, node, multiplierReg,
+         TR::MemoryReference::createWithIndexReg(cg, multiplierAddrReg, tempReg, 16));
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vmuluwm, node, vtmp2Reg, vtmp2Reg, multiplierReg);
+      // restore the multiplier reg
+      generateTrg1MemInstruction(cg, TR::InstOpCode::lxvw4x, node, multiplierReg,
+         TR::MemoryReference::createWithIndexReg(cg, multiplierAddrReg, constant0Reg, 16));
```

Comment:
Although correct, I think you can delay loading the `multiplierReg` initially in line 10535 to not have to load it twice. Currently it's loaded (and not used), then over-written here, then restored, which is redundant.